phoenicis-engines/src/main/java/org/phoenicis/engines/EnginesSource.java

```diff
@@ -49,20 +49,22 @@ public EnginesSource(ScriptInterpreter scriptInterpreter, ObjectMapper objectMap
     public void fetchAvailableEngines(List<CategoryDTO> categoryDTOS, Consumer<List<EngineCategoryDTO>> callback) {
         final InteractiveScriptSession interactiveScriptSession = scriptInterpreter.createInteractiveSession();
 
-        List<EngineCategoryDTO> engines = new ArrayList<>();
+        List<EngineCategoryDTO> engines = Collections.synchronizedList(new ArrayList<>());
         for (CategoryDTO categoryDTO : categoryDTOS) {
             final String engineName = categoryDTO.getName();
             interactiveScriptSession.eval("include([\"Engines\", \"" + engineName + "\", \"Engine\", \"Object\"]);",
                     ignored -> interactiveScriptSession.eval("new " + engineName + "().getAvailableVersions()",
                             output -> {
-                                EngineCategoryDTO wine = new EngineCategoryDTO.Builder().withName(engineName)
-                                        .withDescription(engineName).withSubCategories(unSerialize(output)).build();
-                                engines.add(wine);
-                                callback.accept(engines);
+                                final EngineCategoryDTO engineCategoryDTO = new EngineCategoryDTO.Builder()
+                                        .withName(engineName)
+                                        .withDescription(engineName)
+                                        .withSubCategories(unSerialize(output))
+                                        .build();
+                                engines.add(engineCategoryDTO);
                             }, this::throwError),
                     this::throwError);
         }
-
+        callback.accept(engines);
```

Comment:
Is it always the case that the callback method is called **after** the `interactiveScriptSession` eval calls have been successfully terminated?

Reply:
You are right. It doesn't work like this.

Reply:
Maybe we should add a synchronous way to evaluate a script?
Afterwards we could synchronously execute the scripts and call the callback afterwards.
In addition we could wrap all these steps in a single lambda which we execute in a new thread.

Reply:
I have a different idea. Let me try something.

Reply:
Ok. It works. I'm not sure if there is a better way than calling this function like I did it now.